ALE with finite volume for fluid flow

lindsayad opened the pull request:
## Reason
The AMMT program expects us to do ALE with finite volumes for the fluid flow and CFEM for the displacements

## Design
- Significant development to enable finite volume on displaced mesh/problem
- Add `NodeArg` for functor evaluation
- Implement `ElemArg` for FE for coupling the mesh velocity into the convective term of the momentum and energy transport equations
- Implement Qp evaluations for FV by just passing to `ElemPointArg` overload which does a two term expansion

## Impact
ALE simulation capability with finite volume modeling the NS equations and nodal displacements for the displacing mesh

## Results

This is a two-dimensional melt pool simulation. Coloring is temperature and arrows are pointing in direction of velocity and scaled by the velocity magnitude. A laser spot x-coordinate is a sinusoidal function of time, ratcheting back and forth in the middle half of the domain

https://github.com/idaholab/moose/assets/10248304/8fc770e1-5b16-46ea-875b-dc7793749cfe

AlbertJP7 committed:
Node Argument header functor files

Refs #19420

AlbertJP7 committed:
Changes to AMMT from NodeFunctorArgs

AlbertJP7 committed:
Changes made to NodeFunctorArgs

AlbertJP7 committed:
Edited files for NodeArg to maybe work

AlbertJP7 committed:
Other files

lindsayad committed:
Apply suggestions from code review

Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

lindsayad committed:
clang-format

lindsayad committed:
Fix compilation errors/warnings

lindsayad committed:
Change to use functors in INSAD*DisplaceBoundary objects

lindsayad committed:
Implement NodeArg for MooseVariableFE

lindsayad committed:
Help dumb compilers

lindsayad committed:
Add associated_subdomain to laser input

lindsayad committed:
adding Function functors can be simpler now

lindsayad committed:
Create INSADEnergyMeshConvection

lindsayad committed:
INSADMeshConvection -> INSADMomentumMeshAdvection

lindsayad committed:
INSADEnergyMeshConvection -> INSADEnergyMeshAdvection

lindsayad committed:
Add dot functor material properties to GenericVectorFunctorMaterial

lindsayad committed:
Add mesh convection treatment to FV kernels

To give consistent treatment around material time derivatives I
also had to create new `WCNSFVMassAdvection` and `PWCNSFVMassAdvection`
classes.

lindsayad committed:
Fix bug in 3d-rc-no-slip input

And conduct some other updates to 3d-rc-no-slip

- make w-advection kernel actually apply to w
- remove machine epsilon-ish initial conditions
- coarsen mesh slightly
- use two term boundary expansion
- use skewness correction for all fields

lindsayad committed:
Add use_displaced_mesh = true for remaining FE laser welding mats

This doesn't affect any of the current simulation results but it
could have if we were using `_grad_k` somewhere since although
the temperature values are not functions of the mesh displacement,
the temperature gradients are

lindsayad committed:
Create FVFunctorRadiativeBC

lindsayad committed:
Create FVGaussianEnergyFluxBC

lindsayad committed:
Building our FV ALE input

lindsayad committed:
Fixes to allow FV on displaced meshes

lindsayad committed:
Create INSFVVaporRecoilPressureMomentumFlux

lindsayad committed:
Create functor version of Aria material

lindsayad committed:
Implement some missing variable functor evaluation routines

- ElemArg for finite elements
- ElemQpArg and ElemSideQpArg for finite volumes

lindsayad committed:
Implement evaluateDot with ElemArg for FE

lindsayad committed:
Do separate loops for displaced FV objects

lindsayad committed:
Duplicate functor materials for undisp and disp problems

lindsayad committed:
Make sure to create node functor in PWBBLF

lindsayad committed:
Explicit signed char for AttribDisplaced key

lindsayad committed:
Actually signal an exception in ComputeFVFluxThread

lindsayad committed:
When running displaced, make sure to prepare displaced data

lindsayad committed:
Use structured binding for clarity in buildFiniteVolumeInfo

lindsayad committed:
Make sure displaced elem gets reinit'd for FV elemental kernels

lindsayad committed:
Require qp computations for FVRadiativeHeatFluxBCBase

lindsayad committed:
Constrain away pressure nullspace

lindsayad committed:
Change to creating disp dot functor in the RC UO

lindsayad committed:
Implement FaceArg for MooseVariableFE::evaluateDot

lindsayad committed:
Try evaluating NodeArg for FV with two term expansion

I thought this would better match the sparsity pattern coming out
of the dummy integrated BC (which would also use a two term expansion
to the boundary face) but I'm still getting new nonzero errors

lindsayad committed:
Create 2D FV input

lindsayad committed:
Do all caching of residuals/Jacobians through subproblem

This appears to fix the sparsity pattern issues I was running into
for a reason I have already forgotten

lindsayad committed:
Update elem for mesh adaptivity in FVPointValueConstraint

lindsayad committed:
No need to cacheJacobian; no FV objects use _local_ke

lindsayad committed:
Mark finite volume info as dirty after displacing mesh

lindsayad committed:
Dont scale vels

lindsayad committed:
Move constraint point into middle of domain

lindsayad committed:
Try out INSFVMomentumPressureFlux

lindsayad committed:
Create INSFVMomentumPressureFlux

lindsayad committed:
Disable adaptivity

lindsayad committed:
use fine initial mesh

lindsayad committed:
Decrease solid phase viscosity for less discontinuity

lindsayad committed:
Revert "Try out INSFVMomentumPressureFlux"

This reverts commit 2dcc47a802347bdaf0804ff81e945e57c42c6a8e.

lindsayad committed:
Make sure to setup displaced variables

lindsayad committed:
disp_x shouldn't be required in the interpolator

lindsayad committed:
Futzing around with the 2D FV input

Something still doesn't seem quite right. Convergence is terrible
and I'm not sure the physics look right

lindsayad committed:
Add a great looking 2D finite element input

All the profiles look like I would hope. Most notably the vel-x
points away from the impinging spot whereas for the FV sim the
vel-x is pointing towards the spot. That definitely doesn't seem
right

lindsayad committed:
Add channel flow displaced mesh test

lindsayad committed:
Add displaced lid-driven test

lindsayad committed:
Use one term expansion for FV at nodes

lindsayad committed:
Revert "Use one term expansion for FV at nodes"

This reverts commit d22adab8dc5971efef08e6c0efe0d07dc7536151.

lindsayad committed:
A good result. Hoozaw

lindsayad committed:
Good result on a bit finer mesh

lindsayad committed:
An awesome sim

lindsayad committed:
Remove unimportant lines in FV input

lindsayad committed:
Add 2d-fv test

lindsayad committed:
Avoid recursion in problem-system setups

jiangwen84 commented:
@lindsayad  This looks great. According to your experience, is there any apparent pros of FV version of ALE?

lindsayad committed:
Update gold files for FV evaluated at qps on sides

We now do a two term expansion for FV evaluations at quadrature
points based on the cell center value and gradient whereas previously
we used the "FE" evaluation of an FV variable which just used
the cell center value everywhere in the element

lindsayad commented:
My experience thus far is that it's a lot harder to make it work 😅 It seems less robust; I've had a harder time getting good results for a given set of parameters. For instance, I haven't gotten good results yet in which the "melt pool" starts from a uniformly cold temperature, e.g. 300K, in which the entire medium is initially solid. I was really close to giving up before getting the result shown above. It was pure luck that I switched to an IC in which the material is liquid and observed that I could get some results. So I think there's still a lot to learn and explore

jiangwen84 commented:
> My experience thus far is that it's a lot harder to make it work 😅 It seems less robust; I've had a harder time getting good results for a given set of parameters. For instance, I haven't gotten good results yet in which the "melt pool" starts from a uniformly cold temperature, e.g. 300K, in which the entire medium is initially solid. I was really close to giving up before getting the result shown above. It was pure luck that I switched to an IC in which the material is liquid and observed that I could get some results. So I think there's still a lot to learn and explore

Haha, did not know it was that harder for you :-) I will explore it too. This is really a new and exciting capability. thanks!

lindsayad committed:
Update unit tests

lindsayad committed:
Help dumb compilers

lindsayad committed:
Mark overrides

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1792905/) on b626e8c wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/25574/site)

This comment will be updated on new commits.

lindsayad committed:
Finite volume variables themselves should induce ghosting

Especially with the two term expansion intrinsic to quadrature
point evaluations of finite volume variables, this is very much
necessary

lindsayad committed:
Add missing include in INSADMomentumMeshAdvection.h

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1793779/) on 128b906 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>fe7907</th>
      <th colspan=4>#25574 128b90</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.68%</td>
      <td>85.58%</td>
      <td>-0.10%</td>
      <td>60.73%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>91623</td>
      <td>91711</td>
      <td>+88</td>
      <td>283</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15309</td>
      <td>15453</td>
      <td>+144</td>
      <td>183</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/25574/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/25574/coverage/framework)


## Modules coverage

### Heat conduction

<table>
  <thead>
    <tr>
      <th></th>
      <th>fe7907</th>
      <th colspan=4>#25574 128b90</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>88.46%</td>
      <td>88.01%</td>
      <td>-0.45%</td>
      <td>50.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>3679</td>
      <td>3691</td>
      <td>+12</td>
      <td>23</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>480</td>
      <td>503</td>
      <td>+23</td>
      <td>23</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/25574/coverage_diff/heat_conduction)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/25574/coverage/heat_conduction)

### Navier stokes

<table>
  <thead>
    <tr>
      <th></th>
      <th>fe7907</th>
      <th colspan=4>#25574 128b90</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>84.17%</td>
      <td>84.22%</td>
      <td>+0.05%</td>
      <td>92.65%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>11229</td>
      <td>11338</td>
      <td>+109</td>
      <td>189</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>2112</td>
      <td>2124</td>
      <td>+12</td>
      <td>15</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/25574/coverage_diff/navier_stokes)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/25574/coverage/navier_stokes)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/25574/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/25574/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/25574/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/25574/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/25574/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/25574/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/25574/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/25574/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/25574/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/25574/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/25574/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/25574/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/25574/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/25574/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/25574/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/25574/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/25574/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/25574/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/25574/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/25574/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/25574/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/25574/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/25574/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/25574/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/25574/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/25574/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/25574/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/25574/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 60.73% is less than the suggested 90.0%
- `heat_conduction` new line coverage rate 50.00% is less than the suggested 90.0%



This comment will be updated on new commits.

lindsayad committed:
Skewness correction runs out of AD container space

lindsayad committed:
Remove unused laser welding inputs

lindsayad committed:
Documentation

Apparently class descriptions can only be in the `validParams`
method

GiudGiud commented on framework/contrib/wasp:
dont think this belongs

GiudGiud commented on framework/include/fvkernels/FVPointValueConstraint.h:
```diff
@@ -32,8 +32,10 @@ class FVPointValueConstraint : public FVScalarLagrangeMultiplierConstraint
   static InputParameters validParams();
 
   FVPointValueConstraint(const InputParameters & parameters);
+  virtual void meshChanged() override { setMyElem(); }
```
pressurePin will probably want the same. Does it work better by any chance?

lindsayad replied:
You look for a new element every time you `execute` so `NSFVPressurePin` seems fine as is.

We don't really have any tests of this class with adaptivity, so my only reference point is the new laser welding input ... and yes with this change adaptivity definitely works better ... the singular row is no longer singular

GiudGiud commented on modules/navier_stokes/doc/content/source/kernels/INSADMomentumMeshAdvection.md:
```diff
@@ -0,0 +1,22 @@
+!syntax description /Kernels/INSADMomentumMeshAdvection
+
+`INSADMomentumMeshAdvection` implements the corresponding weak form for the components of
+the term:
```
```suggestion
the term representing the contribution of the mesh displacement to the conservation of momentum:
```

GiudGiud replied:
should there be an INSADMassMeshAdvection? seems like moving the mesh moves mass

GiudGiud commented on modules/heat_conduction/include/bcs/GaussianEnergyFluxBC.h:
```diff
@@ -19,9 +19,13 @@ class GaussianEnergyFluxBC : public ADIntegratedBC
 
   GaussianEnergyFluxBC(const InputParameters & params);
 
+  template <typename T, typename PointType>
```
docstring

GiudGiud commented:
> Implement Qp evaluations for FV by just passing to ElemPointArg overload which does a two term expansion
 
Is this for elem Qp or sideQp ?

GiudGiud commented on framework/include/variables/MooseVariableFE.h:
```diff
@@ -682,22 +686,93 @@ class MooseVariableFE : public MooseVariableField<OutputType>
+  mutable std::vector<GradientType> _current_elem_side_qp_functor_gradient;
+
+  /// The values of the time derivative for the \p _current_elem_side_qp_functor_elem_side
+  mutable std::vector<DotType> _current_elem_side_qp_functor_dot;
```
I kind of wish there was a way to put this code in a separate class, so all the functor stuff lives together.

But a base class would need to still be in the current chain of variable classes to have all the attributes needed
and a derived class would be troublesome in its own way

lindsayad replied:
This kind of caching is unique to the reinit'ing of FE objects so to me this is an ideal place for it

GiudGiud commented on framework/include/variables/MooseVariableFV.h:
```diff
@@ -714,6 +720,30 @@ MooseVariableFV<OutputType>::evaluate(const ElemArg & elem_arg, const StateArg &
+{
+  return (*this)(elem_point.makeElem(), state) +
+         (elem_point.point - elem_point.elem->vertex_average()) *
+             this->gradient(elem_point.makeElem(), state);
```
this is not that intruitie right so let's make it clear in the docs

GiudGiud commented on framework/src/systems/NonlinearSystemBase.C:
```diff
@@ -1544,13 +1546,23 @@ NonlinearSystemBase::computeResidualInternal(const std::set<TagID> & tags)
       FVRange faces(_fe_problem.mesh().ownedFaceInfoBegin(), _fe_problem.mesh().ownedFaceInfoEnd());
       Threads::parallel_reduce(faces, fvr);
     }
+    if (auto displaced_problem = _fe_problem.getDisplacedProblem();
```
add a comment?
we need to do a face-info loop once for undisplaced objects and once for diplaced ones

GiudGiud commented on modules/navier_stokes/include/kernels/INSADMomentumMeshAdvection.h:
```diff
@@ -0,0 +1,68 @@
+  };
+  check_coupled("disp_x");
+  check_coupled("disp_y");
+  check_coupled("disp_z");
```
does this check the block restrictions too

lindsayad replied:
No but the Problem does integrity checks for that

GiudGiud commented on modules/navier_stokes/include/kernels/INSADMomentumMeshAdvection.h:
```diff
@@ -0,0 +1,68 @@
+    // Bypass the UserObjectInterface method because it requires a UserObjectName param which we
+    // don't need
+    auto & obj_tracker = mesh_convection_obj._fe_problem.template getUserObject<INSADObjectTracker>(
+        "ins_ad_object_tracker");
```
should we check block restriction of that one vs this kernel

lindsayad replied:
The tracker is a general user object so it's not block restrictable. It bases its "block restriction" based off the blocks of the INSADMaterials. I already did have an error for this but the message was not very good. I will improve it

GiudGiud commented on modules/navier_stokes/src/bcs/INSADDisplaceBoundaryBC.C:
```diff
@@ -18,17 +18,23 @@ INSADDisplaceBoundaryBC::validParams()
+  params.addRequiredParam<MooseFunctorName>("velocity", "The velocity at which to displace");
   params.addRequiredParam<unsigned short>(
       "component", "What component of velocity/displacement this object is acting on.");
+  params.addRequiredParam<SubdomainName>(
+      "associated_subdomain",
```
this could be limiting in the futue, add a node to split the nodeset if the subdomain dont extend to the entire nodeset?

lindsayad replied:
I'm not quite sure what you mean in your second clause

GiudGiud commented on modules/navier_stokes/src/bcs/INSADDisplaceBoundaryBC.C:
```diff
@@ -39,5 +45,6 @@ INSADDisplaceBoundaryBC::INSADDisplaceBoundaryBC(const InputParameters & paramet
 {
-  return _u - (_u_old + this->_dt * _velocity(_component));
+  const Moose::NodeArg nd{_current_node, _sub_id};
+  return _u - (_u_old + this->_dt * _velocity(nd, determineState())(_component));
```
hardcoding Euler

lets have some check in the constructor?

lindsayad replied:
It's already there 10 lines up per your suggestion on the original PR that added this object :-)

GiudGiud reviewed (commented):
looked over the whole thing. 

it would be good to see a reference solution being matched, for example something pretty contrived in 1D where we impose a displacement and watch the incompressible fluid follow?

or MMS?

lindsayad commented on framework/src/variables/MooseVariableFV.C:
```diff
@@ -51,6 +51,11 @@ MooseVariableFV<OutputType>::validParams()
                                       "Switch that can select between face interpoaltion methods.");
   params.template addParam<bool>(
       "cache_cell_gradients", true, "Whether to cache cell gradients or re-compute them.");
+  // Just evaluating finite volume variables at an arbitrary location in a cell requires a layer of
```
I can change the comment to reflect the fact that there is not a single FV input in our repository that does not require a layer of ghosting

lindsayad commented on modules/navier_stokes/src/fvkernels/INSFVAdvectionKernel.C:
```diff
@@ -32,7 +32,9 @@ INSFVAdvectionKernel::validParams()
-    _rc_vel_provider(getUserObject<INSFVRhieChowInterpolator>("rhie_chow_user_object"))
+    _rc_vel_provider(getUserObject<INSFVRhieChowInterpolator>("rhie_chow_user_object")),
+    _mesh_velocity(isParamValid("mesh_velocity") ? &getFunctor<ADRealVectorValue>("mesh_velocity")
+                                                 : nullptr)
```
This is a remnant of earlier. I've removed it. I will add some new checking for consistency between 'use_displaced_mesh' and displacement parameters in the interpolator object

lindsayad commented on framework/src/variables/MooseVariableFE.C:
```diff
@@ -831,6 +834,442 @@ MooseVariableFE<OutputType>::clearAllDofIndices()
+    }
+  }
+
+  local_soln.resize(n_qp);
+  grad_local_soln.resize(n_qp);
+  if (computing_dot)
+    dot_local_soln.resize(n_qp);
```
For std::vectors, this would only add zeros for new elements

lindsayad commented on framework/src/variables/MooseVariableFE.C:
```diff
@@ -831,6 +834,442 @@ MooseVariableFE<OutputType>::clearAllDofIndices()
+
+  for (const auto qp : make_range(n_qp))
+  {
+    local_soln[qp] = 0;
+    grad_local_soln[qp] = 0;
+    if (computing_dot)
+      dot_local_soln[qp] = 0;
```
This is needed to ensure that all qp-entries are zeroed. The suggestion above would only zero new elements

lindsayad commented on framework/src/variables/MooseVariableFE.C:
```diff
@@ -831,6 +834,442 @@ MooseVariableFE<OutputType>::clearAllDofIndices()
+  }
+  if (cache_eligible)
+    _current_elem_side_qp_functor_elem_side = std::make_pair(elem, side);
+  else
+    _current_elem_side_qp_functor_elem_side = std::make_pair(nullptr, libMesh::invalid_uint);
```
I'll add the comment that I added for the elem-qp version

lindsayad committed:
Add VectorSum postprocessor

lindsayad committed:
Verification testing of 2d-fv.i

lindsayad committed:
Add time derivative of heat capacity

lindsayad committed:
Add back radiation

lindsayad committed:
Revert addition of dcp/dt

lindsayad committed:
Add evaluateDot override to VectorCompositeFunctor

lindsayad committed:
Add INSFVEnergyHeatCapacityGradient

lindsayad committed:
Add missing MOOSE copyright notice

lindsayad committed:
Edit getVelocity doxygen for mesh velocity

lindsayad committed:
Revert "Add INSFVEnergyHeatCapacityGradient"

This reverts commit 8279fa9ec19deea6ddf69c66edac94250fc0a1b1.

lindsayad committed:
Use enthalpy instead of cp*T

lindsayad committed:
MMS testing mesh displacement with FV

lindsayad committed:
Apply mesh velocity on boundary faces as well in RC UO

lindsayad committed:
Trying to get MMS test to work with mesh displacements

It's very tricky trying to compute the MMS time derivative of
displacement with respect to the displaced coordinates when in
the simulation I'm applying the displacement auxiliary kernel with
the reference coordinate

lindsayad committed:
Write non-deprecated syntax in build_hit

lindsayad committed:
Try forcing boundaries since mesh velocity non-zero there

lindsayad committed:
Getting second order convergence with simplified test

So I'm concluding that the mesh displacement treatment is good

lindsayad committed:
Adapt inputs to use enthalpy

lindsayad commented:
> for example something pretty contrived in 1D where we impose a displacement and watch the incompressible fluid follow?
> 
> or MMS?

I've added an MMS case that tests the incorporation of the mesh velocity in the advecting velocity

GiudGiud commented on framework/src/variables/MooseVariableFE.C:
```diff
@@ -831,6 +834,432 @@ MooseVariableFE<OutputType>::clearAllDofIndices()
+              "Our variable should be defined on the requested subdomain ID");
+  const Node & node = *node_arg.node;
+  mooseAssert(node.n_dofs(this->_sys.number(), this->number()),
+              "Our variable must have dofs on the requested node");
```
we could also assert that this dof be a value dof as opposed to a gradient dof

and/or that there be only one dof on the node (per component if vector)

lindsayad replied:
I don't know what a gradient dof is. I don't know of finite element approximate formulations other than u = sum(i) phi_i * u_i

GiudGiud replied:
I think it s a thing in our exotic families, like szabab. This is the kind of thing I just ask Roy about tbh

We essentially do not support them in MOOSE in both the auxkernel and transfer systems

lindsayad committed:
Update python MMS unit test

lindsayad commented on framework/src/postprocessors/VectorSum.C:
```diff
@@ -0,0 +1,52 @@
+VectorSum::VectorSum(const InputParameters & parameters)
+  : GeneralPostprocessor(parameters),
+    _vec(_fe_problem.getNonlinearSystemBase().getVector(
+        _fe_problem.getVectorTagID(getParam<TagName>("vector"))))
```
The communication happens in the `sum()` call on the vector

lindsayad commented on modules/navier_stokes/src/materials/INSFVEnthalpyMaterial.C:
```diff
@@ -31,11 +37,22 @@ INSFVEnthalpyMaterial::INSFVEnthalpyMaterial(const InputParameters & parameters)
+
+  const auto & h = addFunctorProperty<ADReal>(NS::specific_enthalpy,
+                                              [this](const auto & r, const auto & t)
+                                              { return _cp(r, t) * _temperature(r, t); });
```
I amended the doc string for `cp` to say it should be a constant. I do not know how to integrate some arbitrary expression for `_cp` here in this functor. In the Aria material since I have an analytic expression for cp, I implement the analytic integral for cp dtau for h

GiudGiud replied:
oh yeah my bad. 

We should just query _fp directly for h instread of _cp, we should not try to integrate ourselves

GiudGiud replied:
doc string update is fine

lindsayad committed:
Add gradDot functor API

lindsayad committed:
Add missing mesh advection term

This term comes because to convert to the ALE system you have this:

partial phi/partial t -> partial phi/partial t - mesh_vel * grad phi

The last term is *not* the same as -\nablda \cdot (mesh_vel * phi).
So what we end up implementing is:

-mesh_vel * grad_phi -> -(\nabla\cdot(mesh_vel*phi) - phi\nabla\cdot(mesh_vel))

The first term is covered by the flux advection kernel. We now
introduce an elemental kernel to cover the second term

lindsayad committed:
Add momentum mesh advection kernel to MMS test

lindsayad committed:
Add RC to MMS test

lindsayad committed:
Address Guillaume review

- VectorSum -> TagVectorSum
- Class doxygen
- Fix typos
- enthalpy density -> (extensive) enthalpy

lindsayad committed:
Markdown documentation

For
- TagVectorSum
- INSFVMeshAdvection
- INSFVMomentumMeshAdvection

moosebuild commented:
Job [Conda (Intel Mac)](https://civet.inl.gov/job/1812817/) on 25910be : invalidated by @lindsayad

GiudGiud reviewed (dismissed):
functor unit test could use a freshener if you want to get framework coverage up. It s low because the code is tested in modules.

otherwise the new results look good, the mms order in the tests are a good sign and I don't see any issue in the code.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1822647/) on 8276458 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/25574/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/25574/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 3b5aacacf2b2a2166708d2e491b494086b56de9a`

lindsayad committed:
Apply suggestions from code review

- document MooseVariableFV functor overrides
- Comment why we don't cache Jacobians
- Check _face_interp_method to determine skew correction arg
- Comment why we mark finite volume face info dirty
- Comment why we mark _reinit_displaced_elem
- Comment why we do two different face info loops
- Use inverse distance weighting for FV nodal evaluation
- Doxygen for beamFlux
- Add units to input file params
- Fix INSFVFreeSurfaceBC class doxygen
- Remove unused _mesh_velocity from INSFVAdvectionKernel
- Better error message for block restriction in INSADObjectTracker
- Error if disps provided but not on displaced mesh in RC UO
- Comment about why we invalidate cache
- Just index with 0 qp in MooseVariableFE FV overloads
- Fix weird line spacing

Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1822699/) on ecdb631 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/25574/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/25574/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 3b5aacacf2b2a2166708d2e491b494086b56de9a`

GiudGiud commented on modules/navier_stokes/src/userobjects/INSFVRhieChowInterpolatorSegregated.C:
```diff
@@ -155,8 +155,8 @@ VectorValue<ADReal>
-                                                 const THREAD_ID /*tid*/
-) const
+                                                 const THREAD_ID /*tid*/,
+                                                 const bool /*subtract_mesh_velocity*/) const
```
should you assert on subtract_mesh_velocity since it s not implemented

lindsayad replied:
I think I will just error if this object is `_displaced`. It's valid for a "general" postprocessor to call this method with `true` as long as the calculation is not actually displaced

lindsayad committed:
FunctorElementalAux -> FunctorAux allowing nodal evals

lindsayad committed:
Error if user attempts to use segregated UO on displaced mesh

lindsayad committed:
Add analytic node/elem/qp-arg evaluation test

lindsayad committed:
Also test accuracy of grad/dot/grad-dot evaluations

lindsayad commented:
Ok added a meaningful test that should boost framework coverage

lindsayad committed:
Convert to use FunctorAux

moosebuild commented:
Job [Conda (ARM Mac)](https://civet.inl.gov/job/1825615/) on 01b84f4 : invalidated by @lindsayad

GiudGiud commented on modules/navier_stokes/src/materials/INSFVEnthalpyMaterial.C:
```diff
@@ -16,12 +16,18 @@ InputParameters
+  params.addParam<MooseFunctorName>(
+      NS::cp, NS::cp, "The constant value for the specific heat capacity");
+  params.addParam<MooseFunctorName>(
+      NS::enthalpy_density, NS::enthalpy_density, "the name of the enthalpy density");
```
umm it was specific

lindsayad commented:
You'll have to re-approve after the heat transfer merge haha

lindsayad committed:
Unit testing for NodeArg and gradDot

moosebuild commented:
Job [OpenMPI](https://civet.inl.gov/job/1827331/) on 280c09b : invalidated by @lindsayad

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1827334/) on 280c09b wanted to post the following:


The following coverage requirement(s) failed:
- `heat_transfer` coverage rate 88.25% is less than the required 90.0%

lindsayad commented:
All but the Pronghorn failures are unrelated and there's a patch up for that

GiudGiud commented:
Thanks @AlbertJP7 for your contribution on nodal functor arguments! They are actually going to be really important for coupling 1D physics with 3D physics at nodes.

And congrats @lindsayad  on getting this ALE capability in!

lindsayad commented:
Thanks, and thank you for your review as always!
